debian/idraw.1
.TH idraw 1
.SH NAME
idraw \- vector graphics drawing editor
.SH SYNOPSIS
.B idraw 
.I [options] [file]
.SH "DESCRIPTION"
Idraw is a drawing editor that  lets  you  create  and  edit drawings made up of graphics like text, lines, splines, rectangles, polygons, and ellipses.   Drawings  are  stored  in files  that can be printed on a PostScript printer.  You can open an existing drawing when starting up idraw by  typing a file name on the command line.
.PP
Idraw displays a portrait or landscape view of an 8.5 by  11 inch  page in its drawing area.  In a column along the drawing area's left side is a set of  drawing  tool  icons,  and above  the drawing area is a set of pull-down menus containing commands.  A panner in the lower left  corner  lets  you pan  and  zoom the the drawing area.  Along the top is a set of indicators that display editing information.
.SH DRAWING TOOLS
You must engage a tool before you can use it.  You engage  a tool  by  clicking on  its  icon or by typing the character
below and to the right of its icon.  The icon of the drawing
tool  that's  engaged  appears  in  inverted  colors.   Once
engaged, you use the tool by clicking the left mouse  button
in the drawing area.
.PP
The Select, Move, Scale, Stretch, Rotate,  and  Alter  tools
manipulate  existing  graphics.  Magnify makes a part of the
view expand to fill the entire view.  Text, Line, Multiline,
Open  Spline, Ellipse, Rectangle, Polygon, and Closed Spline
create new graphics.  Each tool works as follows:
.TP 15     
Select         
Select a graphic, unselecting all others.   A
graphic  is selected if its handles are visible.  Handles are small inverse-video squares
that either surround the graphic or demarcate
its important points (such as  the  endpoints
of  a  line). If you hold down the shift key,
Select extends the selection: it selects  the
unselected graphic (or unselects the selected
graphic) you clicked on but does not unselect
other  selections.   Clicking  anywhere other
than on a graphic unselects  everything;  you
may also drag a rubberband rectangle around a
group of graphics to select all  of  them  at
once.    Shortcut:  the  right  mouse  button
invokes Select while  the  mouse  is  in  the
drawing area.
.TP
Move           
Move  graphics  from  one  spot  to  another.
Shortcut:  the  middle  mouse  button invokes
Move while the mouse is in the drawing area.
.TP
Scale          
Scale graphics about their centers.
.TP
Stretch        
Stretch graphics vertically  or  horizontally
while tying down the opposite edge.
.TP
Rotate         
Rotate graphics about their centers according
to  the  angle  between  two  radii:  the one
defined by the original  clicking  point  and
the  one  defined  by  the  current  dragging
point.
.TP
Alter          
Modify a graphic's  structure.   This  tool's
effect is described below for each graphic.
.TP
Magnify        
Magnify a portion of the drawing specified by
sweeping  out a rectangular area.  Idraw will
magnify the area to occupy the entire screen,
if possible.
.TP
Text          
Create some text.  Left-click to position the
first  line  of  text,  and then type as much
text as you want.  You  may  use  emacs-style
keystrokes  to edit the text as well as enter
it.  You can leave text editing mode by  typ-
ing ESC or by simply clicking somewhere else.
The Alter tool lets you edit the text  in  an
existing text graphic.
.TP
Line           
Create a line.  The shift key constrains  the
line  to  lie  on  either the vertical or the
horizontal axis.  You may left-click with the
Alter  tool  on  either endpoint of a line to
move the endpoint to a new location.
.TP
Multiline      
Create a set of connected lines.   The  shift
key  constrains each segment to lie on either
the vertical or the  horizontal  axis.   Each
left-click starts a new segment (i.e., adds a
vertex); each right-click  removes  the  last
vertex  added.   The  middle button finalizes
the multiline.  The Alter tool lets you move,
add,  and  remove  vertices  from an existing
multiline.
.TP
Open Spline    
Create an open B-spline.  The shift key  con-
strains  each  control point to lie on either
the vertical or the horizontal axis with  the
preceding  point.   Each  left-click  adds  a
control point; each right-click  removes  the
last  control point added.  The middle button
finalizes the spline.  The  Alter  tool  lets
you move, add, and remove control points from
an existing open spline.
.TP
Ellipse        
Create an ellipse.  The shift key  constrains
the  ellipse  to  the shape of a circle.  The
Alter tool does not affect ellipses.
.TP
Rectangle      
Create a rectangle.  The shift key constrains
the  rectangle to the shape of a square.  The
Alter tool  lets  you  move  the  rectangle's
corners  independently  to  form a four-sided
polygon.
.TP
Polygon        
Create a polygon.  The shift  key  constrains
each  side  to  lie on either the vertical or
the horizontal axis.  Each left-click  starts
a  new  segment  (i.e.,  adds a vertex); each
right-click removes the  last  vertex  added.
The middle button finalizes the polygon.  The
Alter tool lets you  move,  add,  and  remove
vertices from an existing polygon.
.TP
Closed Spline  
Create a closed B-spline.  The shift key con-
strains  each  control point to lie on either
the vertical or the horizontal axis with  the
preceding point.  Each left-click adds a con-
trol point; each right-click removes the last
control   point  added.   The  middle  button
finalizes the spline.  The  Alter  tool  lets
you move, add, and remove control points from
an existing closed spline.

.SH PULL-DOWN MENUS
The pull-down menus File, Edit, Structure, Font, Brush, Pat-
tern,  FgColor,  BgColor,  Align, and View above the drawing
area contain commands for editing the drawing and  for  con-
trolling idraw's execution.  The File menu contains the fol-
lowing commands to operate on files:
.TP
New
Destroy the current drawing  and  replace  it
with an unnamed blank drawing.
.TP
Revert         
Reread the current  drawing,  destroying  any
unsaved changes.
.TP
Open...        
Specify an existing drawing to edit through a
FileChooser(3I),  which  lets  you browse the
file system easily.

Save As...    
Save the current drawing in  a  file  with  a
specific name.
.TP
Save           
Save the current drawing in the file it  came
from.
.TP
Print...       
Send a PostScript version of the drawing to a
printer  or  to a file.  The bold rectangular
outline (called the page boundary)  appearing
in  the drawing area indicates the portion of
the drawing that will appear on  the  printed
page.
.TP
Import Graphic...
Create a graphic from the  information  in  a
file  and insert it into the current drawing.
Idraw can import images  from  files  in  the
following formats: TIFF; PostScript generated
by pgmtops, ppmtops, and idraw; X bitmap for-
mat; and Unidraw format.
.TP
Quit           
Quit idraw.
.PP
The Edit menu contains the following  commands  for  editing
graphics:
.TP
Undo           
Undo the last editing operation.   Successive
Undo  commands undo earlier and earlier edit-
ing operations.
.TP
Redo           
Redo the last editing operation.   Successive
Redo  commands  redo  later and later editing
operations up to the first  operation  undone
by  Undo.   Undone  operations  that have not
been redone are lost as soon as a new  opera-
tion is performed.
.TP
Cut            
Remove the selected graphics from the drawing
and  place  them  in a temporary storage area
called the clipboard.
.TP
Copy           
Copy the selected  graphics  into  the  clip-
board.
.TP
Paste          
Paste copies of the graphics in the clipboard
into  the  drawing.  Together, Cut, Copy, and
Paste let you transfer graphics between draw-
ings  simply  by  cutting graphics out of one
view and pasting them into another.
.TP
Duplicate      
Duplicate the selected graphics and  add  the
copies to the drawing.
.TP
Delete         
Destroy the selected graphics.
.TP
Select All     
Select every graphic in the drawing.
.TP
Flip Horizontal, Flip Vertical
Flip the selected graphics into their  mirror
images along the horizontal or vertical axes.
.TP
90 Clockwise, 90 CounterCW
Rotate  the  selected  graphics  90   degrees
clockwise or counterclockwise.
.TP
Precise Move..., Precise Scale..., Precise Rotate...
Move, scale,  or  rotate  graphics  by  exact
amounts  that  you type in a dialog box.  You
can specify movements in pixels, points, cen-
timeters,  or inches.  Scalings are specified
in terms of magnification factors in the hor-
izontal  and  vertical  dimensions. Rotations
are in degrees.

The Structure menu contains the following commands to modify
the  structure  of  the drawing, that is, the order in which
graphics are drawn:
.TP
Group          
Nest the selected graphics in a newly created               
picture.   A  picture  is just a graphic that               
contains other graphics.  Group allows you to               
build hierarchies of graphics.               
.TP
Ungroup        
Dissolve any selected pictures.
.TP
Bring To Front
Bring the selected graphics to the  front  of
the  drawing so that they are drawn on top of
(after) other graphics.
.TP
Send To Back   
Send the selected graphics to the back of the
drawing   so   that  they  are  drawn  behind
(before) other graphics.
.PP
The Font menu contains a set of fonts in  which  to  display
text.  When you set the current font from the menu, you will
also set all the selected graphics' fonts to that  font.   A
font  indicator  in  the  upper  right  corner  displays the
current font.
.PP
The Brush menu contains a set of brushes with which to  draw
lines.   When  you  set the current brush from the menu, you
will also set all the selected  graphics'  brushes  to  that
brush.   The  nonexistent  brush  draws  invisible lines and
non-outlined graphics.  The arrowhead brushes add arrowheads
to  either  or  both  ends  of  lines,  multilines, and open
splines. A brush indicator in the upper left corner displays
the current brush.
.PP
The Pattern menu contains a set of patterns  with  which  to
fill  graphics but not text.  Text always appears solid, but
you can use a different color than black to get a  halftoned
shade.   When you set the current pattern from the menu, you
will also set all the selected graphics'  patterns  to  that
pattern.   The  nonexistent pattern draws unfilled graphics,
while the other patterns draw graphics filled with a  bitmap
or a halftoned shade.
.PP
The FgColor and BgColor menus contains a set of colors  with
which  to  draw graphics and text.  When you set the current
foreground or background color from the FgColor  or  BgColor
menu,  you  will  also  set all the selected graphics' fore-
ground or background colors.  The ``on'' bits in the bitmaps
for  dashed lines and fill patterns appear in the foreground
color while the ``off'' bits appear in the background color.
A  black  and  white printer will print a halftoned shade of
gray for any color other than black or  white.   The  brush,
pattern, and font indicators all reflect the current colors.
.PP
The Align menu contains  commands  to  align  graphics  with
other  graphics.   The  first  graphic  selected stays fixed
while the  other  graphics  move  in  the  order  they  were
selected  according  to  the  type of alignment chosen.  The
last Align command, Align to Grid, aligns  a  key  point  on
each  selected  graphic to the nearest point on idraw's grid
(see below).
.PP
The View menu contains the following commands:
.TP
New View      
Create a duplicate idraw window containing  a
second  view  of  the  current  drawing.  The
second view may be panned, zoomed, and edited
independently  of  the  first.  Any number of
additional views may be made in this  manner.
Changes  made  to  a drawing through one view
appear synchronously in all  other  views  of
the  same drawing.  You may also view another
drawing in any idraw window via the Open com-
mand.
.TP
Close View     
Close the current idraw window.  Closing  the
last  idraw window is equivalent to issuing a
Quit command.
.TP
Normal Size    
Set the magnification to unity so the drawing
appears at actual size.
.TP
Reduce to Fit  
Reduce the magnification  until  the  drawing
fits entirely within the view.
.TP
Center Page    
Center the view over the center of the 8.5 by
11 inch page.
.TP
Orientation    
Toggle the  drawing's  orientation.   If  the
editor  was  formerly showing a portrait view
of the drawing, it will now show a  landscape
view of the drawing and vice versa.
.TP
Grid on/off    
Toggle idraw's grid on or off.  When the grid
is  on,  idraw draws a grid of equally spaced
points behind the drawing.
.TP
Grid Spacing...
               
Change the grid spacing by specifying one  or
two  values  in  the  units  desired (pixels,
points,  centimeters,  or  inches).   If  two
values  are given (separated by a space), the
first specifies the  horizontal  spacing  and
second  the vertical spacing.  One value will
specify equal horizontal and  vertical  spac-
ing.
.TP
Gravity on/off 
Toggle gravity on or off.  Gravity constrains
tool  operation  to  the grid, whether or not
the grid is visible.

.SH X DEFAULTS
You can customize the number of  undoable  changes  and  the
font, brush, pattern, or color menus by setting resources in
your  X  defaults  database.   Each  string  of   the   form
``idraw.resource:definition'' sets a resource.  For example,
to customize any of the paint menus, set a resource given by
the  concatenation of the menu's name and the entry's number
(e.g., ``idraw.pattern8'') for each entry that you  want  to
override.  All menus use the number 1 for the first entry.

You must set resources only for the entries that you want to
override,  not  all  of them.  If you want to add entries to
the menus, simply set resources for  them.   However,  don't
skip any numbers after the end of the menu, because the menu
will end at the first undefined resource.  To shorten a menu
instead  of  extending  it,  specify  a  blank string as the
resource for the entry following the last.
.PP
Idraw understands the following resources:
.TP
history        
Set the maximum number  of  undoable  changes
               (20 by default).
.TP
initialfont    
Specify the  font  that  will  be  active  on
startup.
 Supply a number that identifies the font  by
its position in the Font menu starting from 1
for the first entry.
.TP
fonti          
Define a custom font to use for the ith entry
in   the   Font  menu.   Give  three  strings
separated by whitespace.   The  first  string
defines  the  font's  name, the second string
the corresponding print font, and  the  third
string   the   print   size.    For  example,
``idraw.font3:8x13bold   Courier-Bold    13''
defines the third font entry.
.TP
initialbrush   
Specify the brush  that  will  be  active  on
startup.   Give  a number that identifies the
brush by  its  position  in  the  Brush  menu
starting from 1 for the first entry.
.TP
brushi         
Define a custom brush  to  use  for  the  ith
entry  in  the  Brush  menu.   The definition
requires two numbers:  a  16-bit  hexadecimal
number to define the brush's line style (each
1 bit draws a dash and each 0 bit produces  a
gap),  and  a  decimal  integer to define the
brush's  width  in  pixels.    For   example,
``idraw.brush2:ffff   1''  defines  a  single
pixel wide  solid  line.  If  the  definition
specifies  only  the string ``none'', then it
defines the nonexistent brush.
.TP
initialpattern 
Specify the pattern that will  be  active  on
startup.   Give  a number that identifies the
pattern by its position in the  Pattern  menu
starting from 1 for the first entry.
.TP
patterni       
Define a custom pattern to use  for  the  ith
entry  in  the Pattern menu.  You can specify
the pattern from a 16x16 bitmap, a  8x8  bit-
map, a 4x4 bitmap, a grayscale number, or the
string ``none''.  You specify the 16x16  bit-
map  with sixteen 16-bit hexadecimal numbers,
the 8x8 bitmap with eight  8-bit  hexadecimal

CHANGES
February 26th 1998  ivtools-0.6.2

Drawing Editor Changes

	* conditionally adds a ConvexHull tool to drawtool if qhull from
 	the University of Minnesota Geometry Center can be found via your
 	PATH environment variable.  To acquire and build this separate
 	executable see
 	http://www.geom.umn.edu/software/download/qhull.html

	* adds a "Compute Attributes" item to the Edit menu
 	(SetAttrByExprCmd).  This brings up a dialog box on the current
 	selection, and allows the user to enter an arbitrary attribute
 	value expression and evaluate it.  For example, if a component had
 	an attribute of "likelihood", a new attribute called
 	"likelihoodsquared" could be set with an expression of
 	"likelihoodsquared=likelihood*likelihood".  Keyboard shortcut in
 	drawtool is "#"
	
Interpreter Changes

	* adds these mathematical functions:
	acos, asin, atan, atan2, cos, sin, tan, pow, exp, log, log10

	* adds a min and max function that promotes operands as needed,
 	similar to the other arithmetic operators.

	* adds a cond operator that returns one of two values based on a
 	boolean test.  Pre-evaluates all parameters (like every other
 	current command except help).  In the future there can be "if",
 	"for", and "while" commands that use a lazy, post-evaluation
 	technique.

	* adds a paste command to comdraw, for use in any derived
 	application that builds up a component that needs pasting in the
 	viewer.

Library Changes

	* TopoFace -- adds methods to construct and return a floating
 	point geomobj that reuses the points without duplicating them:

	    FPointObj* TopoNode::point_obj();
	    FMultiLineObj* TopoEdge::multiline();
	    FFillPolygonObj* TopoFace::polygon()

	* OverlayUnidraw --

	- fixes linear scaling and logarithmic scaling of GrayRaster's

	- fixes mechanism for embedding of a graylevel test strip in
 	GrayRaster's

	- adds an edlauncher() mechanism to OverlayEditor to allow for the
 	inclusion of arbitrary editors into the main.c of one specific
 	editor. 

	- adds a comterplist() mechanism to OverlayEditor to manage a
 	global list of registered interpreters.  Allows for one editor to
 	execute expressions on another editor's associated interpreter.

	- adds more mouse documentation lines to OverlayKit

	- adds OverlayViewer ::ScreenToGraphic and ::GraphicToScreen
 	methods, for converting between the x,y of mouse clicks and the
 	coordinate system of a graphic.  Useful for addressing individual
 	pixels of a raster.

	- adds a boolean OverlayKit::bincheck to test for the presence of
 	an executable before relying on it.  Uses OverlayKit::bintest
 	under the hood.

	- mixes in Observer and Observable as multiply inherited base
 	classes of OverlayComp.  Now other update/notify systems can be
 	constructed around and within the subject/view hierarchy of a
 	Unidraw editor.

	* ComTerp -- 

	- clones the above into a ComFunc::bincheck and ComFunc::bintest,
 	for use of generic ComFunc programmers

	- add a new parameter to ComFunc::argcnts, nargskey (defaulted to
 	zero), which is used by ComTerp to inform the func of how many
 	arguments are associated with keywords.

	- add a ComTerp::stack_height() method to tell the current size of
 	the stack.

Configuration Changes

	* sets up drawserv to link in FrameUnidraw as well as
 	GraphUnidraw, and then tests this with a new Editors menu that
 	allows for the launching of other GraphEditor's or FrameEditor's

	* sets up drawserv to be installed with "make install"

	* sets up comterp to use the original Iterator and UList objects
 	from Unidraw, instead of the ALIterator and AList in the Attribute
 	library.  But then, so raw comterp programs won't have to link in
 	all of Unidraw and InterViews, new libraries called
 	Unidraw-command and IV-common were set up to build with only
 	symbolic links to the necessary files:

	IV-common:  listimpl.c, memory.c, regexp.c, resource.c, textbuffer.c

	Unidraw-common:  iterator.c, ulist.c

	* removes the RCS tags from the TIFF source

Bug Fixes

	* fixes a segfault when attempting to import an empty file

	* fixes bug in the mechanism for managing absolute and relative
 	pathnames for drawtool and related documents.

January 23rd 1998  ivtools-0.6.1

Drawing Editor Changes
	
        * clear any error message on the import, export, open, and save-as
 	dialog boxes from previous attempts upon starting a new attempt.

Interpreter Changes

	* add help command that lists available commands by default, or
 	specific information about each command supplied as an argument.  

	* to support the help command the interpreter was evolved to
 	support a lazy-eval command, one that gets invoked before its
 	arguments get computed.  The next release of ivtools should make
 	use of this with conditional and looping commands, the traditional
 	control constructs of procedural programming.
	
Library Changes

	* TopoFace -- add separate methods to attach start and end nodes to
	  an edge: 
	
	      TopoEdge::attach_start_node(TopoNode*);
 	      TopoEdge::attach_end_node(TopoNode*);

	-- and methods to test if a edge starts or ends at this node:

	      TopoEdge::starts_at(TopoNode*);
	      TopoEdge::ends_at(TopoNode*);

	* OS -- add a method to test the table iterator's current entry.
  	helps avoid a nil pointer reference in case of an empty table:

	      TableEntry(Table)* cur_entry();

	* OverlayUnidraw -- fix bugs caught by ElectricFence, including
 	one to the gs-based selection highlighting mechanism of
 	OverlayUnidraw.  Also fix problem with flushing of
 	deferred-loading rasters and save/restore of arbitrary
	numeric gray-level rasters (GrayRaster)

	* ComTerp -- fix bugs caught by ElectricFence, like referencing
 	off the ComTerp stack.  Increase default buffer size on
 	ComTerpServ to 1k.  Change the ";" operator (SeqFunc) to return
 	the second argument on the ComTerp stack.  Add a
 	ComValue::unkval() to return a static ComValue of UnknownType.


src/glyphs/examples3.1/ips/idraw.cc
/*
 * Copyright (c) 1987, 1988, 1989, 1990, 1991 Stanford University
 * Copyright (c) 1991 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Stanford and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Stanford and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 *
 * IN NO EVENT SHALL STANFORD OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */

/*
 * Read an idraw drawing from a file
 */

#include "figure.h"
#include "idraw.h"
#include <InterViews/box.h>
#include <InterViews/brush.h>
#include <InterViews/character.h>
#include <InterViews/color.h>
#include <InterViews/font.h>
#include <InterViews/place.h>
#include <InterViews/psfont.h>
#include <InterViews/session.h>
#include <InterViews/style.h>
#include <InterViews/tformsetter.h>
#include <InterViews/transformer.h>
#include <OS/file.h>
#include <OS/list.h>
#include <OS/string.h>
#include <ctype.h>

class BrushInfo {
public:
    const Brush* brush_;
    int width_;
    int pattern_;
};

declareList(BrushInfoList,BrushInfo)
implementList(BrushInfoList,BrushInfo)

class Stipple : public Resource {
public:
    Stipple(float dither);

    float dither_;
protected:
    virtual ~Stipple ();
};

Stipple::Stipple (float dither) {
    dither_ = dither;
}

Stipple::~Stipple () { }

declarePtrList(StippleList,Stipple)
implementPtrList(StippleList,Stipple)

static Stipple* no_stipple = (Stipple*) -1;

struct FigureInfo {
    const char* name;
    boolean brush;
    boolean foreground;
    boolean background;
    boolean font;
    boolean pattern;
    boolean transformer;
    int coords;
    int skip;
};
    
static FigureInfo early_info[] = {
    { "Idraw", 1, 1, 1, 1, 1, 1, 0, 0 },
    { "BSpl",  1, 1, 1, 0, 1, 1, -1, 0 },
    { "Circ",  1, 1, 1, 0, 1, 1, 1, 0 },
    { "CBSpl", 1, 1, 1, 0, 1, 1, -1, 0 },
    { "Elli",  1, 1, 1, 0, 1, 1, 1, 0 },
    { "Line",  1, 1, 1, 0, 1, 1, 2, 0 },
    { "MLine", 1, 1, 1, 0, 1, 1, -1, 0 },
    { "Pict",  1, 1, 1, 1, 1, 1, 0, 0 },
    { "Poly",  1, 1, 1, 0, 1, 1, -1, 0 },
    { "Rect",  1, 1, 1, 0, 1, 1, 2, 0 },
    { "Text", 0, 1, 0, 1, 0, 1, 0, 0 },
    { "eop",  0, 0, 0, 0, 0, 0, 0, 0 },
    {nil, 0, 0, 0, 0, 0, 0, 0}
};

static FigureInfo version_10_info[] = {
    { "Idraw", 1, 1, 1, 1, 1, 1, 0, 0 },
    { "BSpl",  1, 1, 1, 0, 1, 1, -1, 1 },
    { "Circ",  1, 1, 1, 0, 1, 1, 1, 0 },
    { "CBSpl", 1, 1, 1, 0, 1, 1, -1, 0 },
    { "Elli",  1, 1, 1, 0, 1, 1, 1, 0 },
    { "Line",  1, 1, 1, 0, 1, 1, 2, 1 },
    { "MLine", 1, 1, 1, 0, 1, 1, -1, 1 },
    { "Pict",  1, 1, 1, 1, 1, 1, 0, 0 },
    { "Poly",  1, 1, 1, 0, 1, 1, -1, 0 },
    { "Rect",  1, 1, 1, 0, 1, 1, 2, 0 },
    { "Text", 0, 1, 0, 1, 0, 1, 0, 0 },
    { "eop",  0, 0, 0, 0, 0, 0, 0, 0 },
    { nil, 0, 0, 0, 0, 0, 0, 0, 0 }
};

class IdrawReaderImpl {
    friend class IdrawReader;

    InputFile* file_;
    int version_;
    FigureInfo* figinfo_;
    const char* start_;
    const char* end_;
    const char* cur_;
    BrushInfoList brushes_;
    StippleList stipples_;
    static Brush* no_brush;

    Glyph* load(
	const Brush* pb, const Color* pfg, const Color* pbg,
	const Font* pf, const Stipple* ps
    );
    boolean fill();
    boolean read(String&);
    boolean read(char&);
    boolean read(int&);
    boolean read(float&);
    const Brush* read_brush();
    const Color* read_color();
    const Font* read_font();
    const Stipple* read_stipple();
    void read_transformer(Transformer&);
    const Color* dither_color(const Color* fg, const Color* bg, float dither);
    void skip();
};

Brush* IdrawReaderImpl::no_brush = (Brush*)-1;

Glyph* IdrawReader::load(InputFile* f) {
    IdrawReaderImpl impl_;
    impl_.file_ = f;
    impl_.start_ = nil;
    impl_.end_ = nil;
    impl_.cur_ = nil;
    return impl_.load(nil, nil, nil, nil, nil);
}

Glyph* IdrawReaderImpl::load(
    const Brush* pb, const Color* pfg, const Color* pbg,
    const Font* pf, const Stipple* ps
) {
    const LayoutKit& layout = *LayoutKit::instance();
    skip();
    Transformer tx;
    Glyph* glyph = nil;
    String name;
    if (read(name)) {
	if (name == "Idraw") {
	    read(version_);
	    figinfo_ = version_ < 10 ? early_info : version_10_info;
	}
	FigureInfo* fig;
	for (fig = &figinfo_[0]; fig->name != nil; fig++) {
	    if (name == fig->name) {
		break;
	    }
	}
        const Brush* b = fig->brush ? read_brush() : nil;
        const Color* fg = fig->foreground ? read_color() : nil;
        const Color* bg = fig->background ? read_color() : nil;
        const Font* f = fig->font ? read_font() : nil;
        const Stipple* s = fig->pattern ? read_stipple() : nil;
        if (fig->transformer) {
            read_transformer(tx);
        }

        if (pb) {
	    b = pb;
	}
        if (pfg) {
	    fg = pfg;
	}
        if (pbg) {
	    bg = pbg;
	}
        if (pf) {
	    f = pf;
	}
        if (ps) {
	    s = ps;
	}

	String figname(fig->name);
        if (fig->name == nil) {
            ; // error
        } else if (figname == "Idraw" || figname == "Pict") {
            Glyph* pic = layout.overlay();
	    for (;;) {
                Glyph* g = load(b, fg, bg, f, s);
		if (g == nil) {
		    break;
		}
		pic->append(g);
            }
            glyph = pic;
        } else if (figname == "eop") {
            glyph = nil;
        } else if (figname == "Text") {
            skip();
	    String s;
	    read(s);
	    char ch;
	    read(ch);
            Glyph* col = layout.vbox();
            Glyph* line = layout.hbox();
	    FontBoundingBox bbox;
	    f->font_bbox(bbox);
            Coord lineheight = bbox.font_ascent() + bbox.font_descent();
	    while (read(ch) && ch != ']') {
                if (ch == '\n') {
                    line->append(layout.strut(f));
                    col->append(
			layout.fixed_span_dimension(
			    line, Dimension_Y, lineheight
			)
		    );
                    line = layout.hbox();
                } else if (ch == ' ') {
		    line->append(new Character(' ', f, fg));
                } else if (ch != ')' && ch != '(') {
                    if (ch == '\\') {
			read(ch);
			if (isdigit(ch)) {
			    ch -= '0';
			    ch *= 8;
			    char digit;
			    read(digit);
			    ch = (ch * 8) + digit - '0';
			    read(digit);
			    ch = (ch * 8) + digit - '0';
			}
                    }
                    line->append(new Character(ch, f, fg));
                }
            }
            Transformer fixtext;
            fixtext.translate(0, bbox.font_descent() - lineheight);
            glyph = new TransformSetter(col, fixtext);
        } else {
            skip();
            int c = fig->coords;
            if (c == -1) { 
		read(c);
            }
            Coord* x = new Coord[c];
            Coord* y = new Coord[c];
            float xx, yy;
            for (int i = 0; i < c; ++i) {
		read(xx);
		read(yy);
		x[i] = xx;
		y[i] = yy;
            }

            const Brush* brush = (b != no_brush) ? b : nil;
            const Color* stroke = fg;
            const Color* fill = (
                (s != no_stipple) ? dither_color(fg, bg, s->dither_) : nil
            );
            if (figname == "Line") {
                glyph = new Line(brush, stroke, fill, x[0], y[0], x[1], y[1]);
            } else if (figname == "BSpl") {
                glyph = new Open_BSpline(brush, stroke, fill, x, y, c);
            } else if (figname == "CBSpl") {
                glyph = new Closed_BSpline(brush, stroke, fill, x, y, c);
            } else if (figname == "MLine") {
                glyph = new Polyline(brush, stroke, fill, x, y, c);
            } else if (figname == "Poly") {
                glyph = new Polygon(brush, stroke, fill, x, y, c);
            } else if (figname == "Rect") {
                glyph = new Rectangle(
		    brush, stroke, fill, x[0], y[0], x[1], y[1]
		);
            } else if (figname == "Circ") {
		Coord radius;
		read(radius);
                glyph = new Circle(brush, stroke, fill, x[0], y[0], radius);
            } else if (figname == "Elli") {
		Coord r1, r2;
		read(r1);
		read(r2);
                glyph = new Ellipse(brush, stroke, fill, x[0], y[0], r1, r2);
            } else {
                glyph = nil;
            }
            delete [] x;
            delete [] y;
        }
        for (int extra = fig->skip; extra > 0; --extra) {
            skip();
        }
    }
    if (glyph != nil && !tx.identity()) {
	glyph = new TransformSetter(glyph, tx);
    }
    return glyph;
}

boolean IdrawReaderImpl::fill() {
    if (cur_ >= end_) {
	int n = file_->read(start_);
	if (n <= 0) {
	    return false;
	}
	cur_ = start_;
	end_ = start_ + n;
    }
    return true;
}

boolean IdrawReaderImpl::read(String& s) {
    if (!fill()) {
	return false;
    }
    const char* p1 = cur_;
    while (p1 < end_ && isspace(*p1)) {
	++p1;
    }
    const char* p2 = p1;
    while (p2 < end_ && !isspace(*p2)) {
	++p2;
    }
    cur_ = p2;
    s = String(p1, p2 - p1);
    return true;
}

boolean IdrawReaderImpl::read(char& c) {
    if (!fill()) {
	return false;
    }
    c = *cur_++;
    return true;
}

boolean IdrawReaderImpl::read(int& i) {
    String s;
    return read(s) && s.convert(i);
}

boolean IdrawReaderImpl::read(float& f) {
    String s;
    return read(s) && s.convert(f);
}

void IdrawReaderImpl::skip() {
    String s;
    while (read(s) && s != "%I") { }
}

const Brush* IdrawReaderImpl::read_brush() {
    skip();
    String s;
    read(s);
    read(s);
    if (s == "u") {
        return nil;
    }
    if (s == "n") {
        return no_brush;
    }
    int pattern;
    int width;
    s.convert(pattern);
    read(width);

    for (ListItr(BrushInfoList) i(brushes_); i.more(); i.next()) {
	BrushInfo& b = i.cur_ref();
	if (b.width_ == width && b.pattern_ == pattern) {
	    return b.brush_;
	}
    }
    BrushInfo* b = new BrushInfo;
    b->brush_ = new Brush(pattern, Coord(width));
    Resource::ref(b->brush_);
    b->width_ = width;
    b->pattern_ = pattern;
    brushes_.append(*b);
    return b->brush_;
}

const Color* IdrawReaderImpl::read_color() {
    skip();
    String s;
    read(s);
    read(s);
    if (s == "u") {
	return nil;
    }
    float r, g, b;
    read(r);
    read(g);
    read(b);
    return new Color(r, g, b, 1.0);
}

const Color* IdrawReaderImpl::dither_color(
    const Color* f, const Color* b, float dither
) {
    float fr, fg, fb;
    float br, bg, bb;
    f->intensities(fr, fg, fb);
    b->intensities(br, bg, bb);
    float red = (1 - dither) * fr + dither * br;
    float green = (1 - dither) * fg + dither * bg;
    float blue = (1 - dither) * fb + dither * bb;
    return new Color(red, green, blue);
}

const Font* IdrawReaderImpl::read_font() {
    skip();
    String s;
    read(s);
    read(s);
    if (s == "u") {
	return nil;
    }
    String psname, ptsize;
    read(psname);
    NullTerminatedString psname_nt(psname);
    float pointsize;
    read(pointsize);
    if (PSFont31::exists(psname_nt.string())) {
	NullTerminatedString s_nt(s);
	return new PSFont31(psname_nt.string(), pointsize, s_nt.string(), 1.0);
    }
    String default_font("fixed");
    Session::instance()->style()->find_attribute("font", default_font);
    return Font::lookup(default_font);
}

const Stipple* IdrawReaderImpl::read_stipple() {
    skip();
    String s;
    read(s);
    read(s);
    if (s == "u") {
	return nil;
    }
    if (s == "n") {
        return no_stipple;
    }
    float dither;
    s.convert(dither);
    for (ListItr(StippleList) i(stipples_); i.more(); i.next()) {
	Stipple* st = i.cur();
	if (st->dither_ == dither) {
	    return st;
	}
    }
    Stipple* st = new Stipple(dither);
    Resource::ref(st);
    stipples_.append(st);
    return st;
}

void IdrawReaderImpl::read_transformer(Transformer& t) {
    skip();
    String s;
    read(s);
    read(s);
    if (s != "u") {
	float a[6];
	for (int i = 0; i < 6; i++) {
	    read(a[i]);
	}
        Transformer tt(a[0], a[1], a[2], a[3], a[4], a[5]);
        t = tt;
    }
}
